### PR TITLE
Model management proposal

### LINKED ISSUE #139: Create a proposal for managing model weights and adapters
Take into account:

* #3 
* #132 
* #138

### PR DESCRIPTION
**Pull request by nstogner**
Fixes #139

### CODE REVIEW

File: proposals/model-management.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  # While standard storage profile names are recommended, they should be open ended
+  # to allow for multiple different admin-defined varieties:
+  MyCustomBucket:
+    bucket:
+    # ...
+```
+
+### Bucket storage
+
+This mode relies on access to a bucket. To support zero external dependencies and local deployment this bucket could be provided by a Helm-deployed Minio (possible License issue?) / [SeaweedFS](https://github.com/seaweedfs/seaweedfs) (mature/secure?) instance.
```
Comment (samos123) [inline proposals/model-management.md:54]: Seaweed would be my choice. I looked into Seaweedfs in the past for same use case. Seaweed is lightweight and has a single binary option.

File: proposals/model-management.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+* KubeAI should be able to be run on clouds with heterogenous disk features (limited support for ReadOnlyMany)
+
+## Solution
+
+KubeAI should support configurable storage for both models and adapters. These settings should be configurable at the `kind: Model` level as some variants (like GCP Hyperdisk ML) are expensive and generally only recommended for larger models.
+
+```yaml
+kind: Model
+spec:
+  storageProfile:
```
Comment (samos123) [inline proposals/model-management.md:24]: is StorageProfile optional? Can we just leave it out and let it pull from HF?

File: proposals/model-management.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    bucket:
+      url: s3://...
+      # ...
+  ReadWriteManyDisk:
+    disk:
+      storageClass: aws-efs
+      # ...
+  ReadManyDisk:
+    disk:
+      storageClass: gcp-hyperdisk
```
Comment (samos123) [inline proposals/model-management.md:43]: so would this be a single hyperdisk shared with all models? and each model is in a subdirectory?

File: proposals/model-management.md (hunk 2 above)
Comment (nstogner) [reply proposals/model-management.md:27]: For pulling models, yes, thinking we have a storage profile option of `None`. I added a block for default storage profiles when none are specified on the Model.

File: proposals/model-management.md (hunk 3 above)
Comment (nstogner) [reply proposals/model-management.md:52]: Updated the proposal a little more - ML Hyperdisk would be mounted once to populate it with a single model, then mounted read-only by multiple servers.

File: proposals/model-management.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    adapters: Bucket
+```
+
+Storage profiles should be defined at the KubeAI config.yaml level:
+
+```yaml
+# Set on Model objects if .spec.storageProfile is undefined.
+defaultStorageProfiles:
+  model: None # None is a special storage mode where each server pulls the model from source.
+  adapters: Bucket
```
Comment (samos123) [inline proposals/model-management.md:38]: Maybe adapters should be None too and then it pulls from HF?

I am thinking we have a default where there is no requirement for any buckets, not even locally within the cluster.

Comment (nstogner) [reply proposals/model-management.md:38]: I was thinking about the situation where we support finetuning (a job that KubeAI manages the produces LoRA adapters). I was thinking we would need a storage mechanism not just a retrieval strategy. However you made me think we could just push those back up to huggingface with the token they already configured.

Comment (nstogner) [reply proposals/model-management.md]: Updated

**Comment by samos123** [comment]
I think I get the proposal now! So would it be possible to express this as well once we support GCS/S3 as a model location?
```
kind: Model
spec:
  url: "gcs://my-bucket/Meta-Llama-3.1-7b-full-training"
```

The use case would be pushing full merged models to a specific location. Or some private model that uses llama 3.1 7b architecture but was trained from scratch.

File: proposals/model-management.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+  - name: sarcasm
+    url: "hf://myrepo/sarcastic-lora-adapter-for-llama-3-8b-Instruct"
+
+# Model that explicitly asks for a cache profile
+kind: Model
+spec:
+  url: "hf://meta-llama/Meta-Llama-3.1-405B-Instruct"
+  cacheProfile: HyperDisk
+
+# A Model without any cache specified will use the default set in KubeAI's config.yaml
```
Comment (samos123) [inline proposals/model-management.md:42]: and the default cacheProfile on installation is cacheProfile: None?

File: proposals/model-management.md [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+If a user submits a finetuning request (`POST /v1/fine_tuning/jobs`), they have the option to specify a location to store the adapters:
+
+```json
+# OpenAI-compatible finetune job request
+{
+  # Model.metadata.name
+  "model": "llama-3.1-8b",
+  # KubeAI-Only: Training file could be specified as URL: "s3://..."
+  "training_file": "<id-of-file-uploaded-via-files-api>",
```
Comment (samos123) [inline proposals/model-management.md:64]: +1 to using something like s3:// or http://

File: proposals/model-management.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+Model cache profiles and finetune default storage locations should be defined at the KubeAI `config.yaml` level:
+
+```yaml
+finetuning:
+  storage:
+    defaultURLTemplate: "s3://my-bucket/finetuning/{model}/{job}"
+  
+models:
+  cache:
+    defaultProfile: Bucket # Could also be "None"
```
Comment (samos123) [inline proposals/model-management.md:97]: if not specified I'm assuming the helm default values would look like this:
```
models:
  cache:
    defaultProfile: ""
    profiles: {}
```

File: proposals/model-management.md [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+  - Typically provisioned outside of the k8s API
+
+##### Static Volume Storage
+
+This mode relies on a Kubernetes PersistentVolumes for storage.
+
+- How it works: mounted as ReadWriteOnce to initialize, then ReadOnlyMany for serving
+- Examples: GCP Hyperdisk ML
+- NOTEs:
+  - Only supported for models, NOT adapters!
```
Comment (samos123) [inline proposals/model-management.md:142]: I don't get why this wouldn't be supported for adapters?

File: proposals/model-management.md [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+- NOTEs:
+  - Works for models and adapaters.
+  - Typically provisioned outside of the k8s API
+
+##### Static Volume Storage
+
+This mode relies on a Kubernetes PersistentVolumes for storage.
+
+- How it works: mounted as ReadWriteOnce to initialize, then ReadOnlyMany for serving
+- Examples: GCP Hyperdisk ML
```
Comment (samos123) [inline proposals/model-management.md:140]: GCP Persistent Disk ReadManyOnly is another option

File: proposals/model-management.md (hunk 7 above)
Comment (nstogner) [reply proposals/model-management.md:97]: Maybe more explicit with None:

```yaml
models:
  cache:
    defaultProfile: None
    profiles: {}
```

File: proposals/model-management.md (hunk 8 above)
Comment (nstogner) [reply proposals/model-management.md:142]: Disks like Hyperdisk ML (and other PDs in GCP) can be mounted as ReadWriteOnce or ReadOnlyMany ... Supporting this type of disk for adapters is complicated:
* A finetune Job needs to output adapters by mounting a new disk in RWO mode (doable - but forces 1:1 adapter-to-disk which is probably not justified based on adapter's being small in size)
* Model servers need to be able to mount these adapter-specific disks dynamically to swap the adapter that is being used for a given server like vLLM  that is serving a base model - see #132 (I dont know of a way of doing tackling this as volumes specified in a Pod spec are not dynamic - require Pod recreation)

*...Will add these details to proposal*

File: proposals/model-management.md (hunk 9 above)
Comment (nstogner) [reply proposals/model-management.md:140]: Good callout - will add that in but mention the max-attachment limitation

File: proposals/model-management.md (hunk 5 above)
Comment (nstogner) [reply proposals/model-management.md:42]: yep!

Comment (BOsterbuhr) [reply proposals/model-management.md:42]: When you say "will use the default set in KubeAI's config.yaml" does that mean the model catalog at `charts/kubeai/charts/models/values.yaml`? If so, and a model is used that isn't in the model catalog, will cacheProfile be a required field?

**Review by samos123 — APPROVED** [review]

**Comment by michaelfeil** [comment]
@nstogner Feedback - potentially the similar/samegoes for training data. Training data usually exceeds request size, and duration exceeds timeout. S3 might be a better fit wrt to persistence.

**Pull request merged**